### ISSUE TITLE
Need an official way to open a group (or run context menu items)

### ISSUE THREAD

**Issue by arigo** [SketchUp/api-issue-tracker#129]
# Feature Request

In our extension "VR Sketch", the user in VR can edit the model, including opening and closing groups and components.  When the user in VR chooses to close a group, we call ``active_model.close_active``.  However, when she chooses to open a group, we have no official way to perform the operation in the SketchUp API.

We are currently using a gross and Windows-only hack, using a mixture of approaches, which I paste below in case somebody else is interested.

Maybe a more generally useful solution would be to let ``Sketchup.send_action(id)`` trigger context menu items.  The code below is a hack to dig out the id corresponding to the "Edit group" context menu.  This id is variable, so we need to emulate the user clicking through the menu until he reaches the right menu command, and read the current id.  I guess that the id is variable because the context menu is itself variable, and any number of extensions can add stuff to it.  Still, if "edit group" and most of the other standard items had a fixed and public id, it would solve the problem.

```#! ruby
require 'fiddle'
require 'fiddle/types'
require 'fiddle/import'
#...

        module Kernel32
            extend Fiddle::Importer
            dlload 'kernel32.dll'
            include Fiddle::Win32Types
            extern 'DWORD GetCurrentThreadId()'
        end

        module User32
            extend Fiddle::Importer
            dlload 'user32.dll'
            include Fiddle::Win32Types
            extern 'HWND GetAncestor(HWND, UINT)'
            #extern 'BOOL EnumThreadWindows(DWORD, WNDENUMPROC, LPARAM)'
            extern 'BOOL EnumThreadWindows(DWORD, PVOID, PVOID)'
        end

        def hack_prepare_open_group()
            # call this once, e.g. when your extension loads
            require 'Win32API'

            @hack_GetMenu = Win32API.new("user32.dll", "GetMenu", ['L'], 'L')
            @hack_GetMenuItemCount = Win32API.new("user32.dll", "GetMenuItemCount", ['L'], 'I')
            @hack_GetMenuItemID = Win32API.new("user32.dll", "GetMenuItemID", ['L', 'I'], 'I')
            @hack_GetSubMenu = Win32API.new("user32.dll", "GetSubMenu", ['L', 'I'], 'L')
            @hack_SendMessage = Win32API.new("user32.dll", "SendMessage", ['L', 'I', 'L', 'L'], 'L')

            # the fiddle parts are based on https://github.com/SketchUp/testup-2/blob/master/src/testup/win32.rb
            thread_id = Kernel32.GetCurrentThreadId()
            main_hwnd = 0

            enumWindowsProc = Fiddle::Closure::BlockCaller.new(Fiddle::TYPE_INT,
              [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP]) { |hwnd, lparam|
                hwnd = User32.GetAncestor(hwnd, 3)   # GA_ROOTOWNER
                hmenu = @hack_GetMenu.call(hwnd)
                next 1 if hmenu == 0
                heditmenu = @hack_GetSubMenu.call(hmenu, 1)
                next 1 if heditmenu == 0
                menulength = @hack_GetMenuItemCount.call(heditmenu)
                contextmenu = @hack_GetSubMenu.call(heditmenu, menulength - 1)
                next 1 if contextmenu == 0
                
                main_hwnd = hwnd
                next 0
            }
            User32.EnumThreadWindows(thread_id, enumWindowsProc, 0)

            UI.messagebox("Cannot get the top-level SketchUp window's Edit menu.  Opening groups in VR will not work") if main_hwnd == 0
            @hack_hwnd = main_hwnd
        end

        def hack_open_group(model, grp)
            if (grp.is_a? Sketchup::Group) || (grp.is_a? Sketchup::ComponentInstance)
                found = 0
                model.active_entities.each { |ent|
                    if ent == grp
                        found += 1
                    end
                }
                if found > 0
                    Sketchup.active_model.selection.clear
                    Sketchup.active_model.selection.add(grp)

                    hwnd = @hack_hwnd
                    hmenu = @hack_GetMenu.call(hwnd)
                    @hack_SendMessage.call(hwnd, 0x0116, hmenu, 0)
                    heditmenu = @hack_GetSubMenu.call(hmenu, 1)
                    @hack_SendMessage.call(hwnd, 0x0117, heditmenu, 1)
                    menulength = @hack_GetMenuItemCount.call(heditmenu)
                    contextmenu = @hack_GetSubMenu.call(heditmenu, menulength - 1)
                    id = @hack_GetMenuItemID.call(contextmenu, 0)
                    Sketchup.send_action(id)
                else
                    puts("open_close_group: group not in active_entities")
                end
            else
                puts("open_close_group: not a group id")
            end
        end
```

**Comment by DanRathbun** [comment]
> The code below is a hack to dig out the id corresponding to the "Edit group" context menu. This id is variable, so we need to emulate the user clicking through the menu until he reaches the right menu command, and read the current id.

a. You would need to have a group or component selected in order for SketchUp to create a context menu for you to iterate.

b. Also, I think you'd need to know the exact menu item text in ALL the languages that SketchUp is built in, to choose the correct item.

**Comment by arigo** [comment]
Yes:

a. That's what the code above does

b. I specifically didn't suggest anything based on the menu item text.  In the code above I pick the first menu item, which is probably always "Edit group".

**Comment by thomthom** [comment]
Internal issue: SU-39162

**Comment by Eneroth3** [comment]
A `ModeL#active_path=` method would be a nice way to directly access any drawing context, without having to dig down one step at a time.

Had `Model#active_path` done the reasonable thing and returned empty array, not nil, in the top drawing context, `active_path += [instance]` could be used as a simple clean way to open an instance in the active drawing context.

**Comment by DanRathbun** [comment]
> @arigo, b. I specifically didn't suggest anything based on the menu item text. In the code above I pick the first menu item, which is probably always "Edit group".

(Just noticed ...) In all my installs on Windows it seems to occupy the **FIFTH** position, not the first.

Also, extensions can add items to the context menu, and at least on Windows can insert items into the menu item list at any position, causing the "Edit Group" item to change position. So relying upon a certain menu position is a fragile pattern.

----

FYI, I have noticed quite a few requests in the public forums over the years for an API method to open an edit context in the GUI.

Should such a command do a **zoom selection** upon the context's entities ?

Ie, imagine if an extension changes the edit context to one that is off screen ... which would be confusing for the user.

**Comment by thomthom** [comment]
> Should such a command do a zoom selection upon the context's entities ?

I don't think so, it would make the API method non-orthogonal.

> Had Model#active_path done the reasonable thing and returned empty array,

Indeed - empty array would have been better. There are several places in the API where `nil` wasn't the best return value. Sandy Metz' article on [Suspicions of nil](https://www.sandimetz.com/blog/2014/12/19/suspicions-of-nil) still rings strong with me.

**Comment by DanRathbun** [comment]
>Should such a command do a zoom selection upon the context's entities ?
>
>> I don't think so, it would make the API method non-orthogonal.

Okay, **agreed**. 

Make a note that the doc should suggest the coder be sure to bring the context into view if it is not. (Perhaps show the use of [`View#zoom(group)`](http://ruby.sketchup.com/Sketchup/View.html#zoom-instance_method) in the method's example. And that it should be adjusted so that the dotted context box is also in view to let the user know they are within the group's edit context.)

**Comment by arigo** [comment]
Thanks for the addition!

@DanRathbun: ah, it's in the fifth position in the right-button-click menu for me too, but the first in the "Edit" -> "Component (N in model)" submenu.  The right-button-click menu starts with four extra items.

Can you tell me how an extension would add a menu item somewhere else than at the end?  I've only seen Menu::add_item()...

**Comment by Eneroth3** [comment]
> Can you tell me how an extension would add a menu item somewhere else than at the end? I've only seen Menu::add_item()...

add_item takes a second, undocumented, index parameter. I've used on rare occasions to get File > Open Newer Version to show up next to File > Open, File > Save in Legacy Format next to File > Save and Context > Flip Individually Along next to Flip Along. It is however something I'd recommend against unless the new action has a very tight relation to an existing one.

Regarding changing drawing context I don't see any reason to connect it to the camera, not even in the documentation. In some cases I'm sure it's valid to do both actions at the same time, but they are two separate actions.

**Comment by DanRathbun** [comment]
Then perhaps you can imagine how the user would react when the edit context is changed leaving the user in control but the context is way off screen so the user doesn't actually KNOW that the context has changed and cannot see what it is they are about to edit ?

**Comment by DanRathbun** [comment]
> @arigo 
Can you tell me how an extension would add a menu item somewhere else than at the end? I've only seen Menu::add_item()...

This ...
```ruby
UI.add_context_menu_handler {|menu| menu.add_item("Test",0) { puts "okay" } }
```
... will defeat your workaround by moving the location of the "Edit" menu item.

**Comment by Eneroth3** [comment]
> Then perhaps you can imagine how the user would react when the edit context is changed leaving the user in control but the context is way off screen so the user doesn't actually KNOW that the context has changed and cannot see what it is they are about to edit ?

It's not like you set up a timer than makes the edit context change at random times. It is the user that changes the edit context one way or another. If an extension changes the edit context in a way that is unexpected or isn't transparent, then that is the problem.

The description for Camera#perspective= doesn't mention how to go to a standard view, even if you sometimes want to do both, and Entities#add_face doesn't have a note for what method to use to add edges, even if you may also want to add edges while you are already drawing stuff. commit_operation doesn't say how to trigger undo and undo doesn't say how to trigger redo.

**Comment by DanRathbun** [comment]
> If an extension changes the edit context in a way that is unexpected or isn't transparent, then that is the problem.

Correct. This is what my suggestion is all about. A simple warning that coders should use this API feature (IF it ever gets implemented,) in a way that is NOT transparent to the user. IE, if you are to leave the user in an **new** edit context (instead of using it in a temporary way,) ... then make it apparent to the user, such as being sure the edit context is displayed in the viewport.

It does not hurt to give best advice. There are so many past complaints about API knowledge that is not documented in the API docs that I've lost count of the complaints. Just because you can point to examples of terse examples in the API doesn't mean that this should continue to be the norm.

If it is felt that the specific method YARDoc entries should be kept as concise as possible, then there still can be **See Also** links into supplemental files on "Best Practices for SketchUp Extension GUI Use" or similar. Like what we had discussed for a primer on use of Observers or other topics like Animation, etc.

**Comment by DanRathbun** [comment]
> In our extension "VR Sketch", the user in VR can edit the model, including opening and closing groups and components. When the user in VR chooses to close a group, we call `active_model.close_active`. However, when she chooses to open a group, we have no official way to perform the operation in the SketchUp API.

@arigo Okay .. now that SketchUp 2020 has added the following ...
http://ruby.sketchup.com/Sketchup/Model.html#active_path=-instance_method
#### Open an instance
```ruby
model = Sketchup.active_model
instance = model.active_entities.grep(Sketchup::ComponentInstance).first
instance_path = Sketchup::InstancePath.new([instance])
model.active_path = instance_path
```
#### Close all instances
```ruby
model = Sketchup.active_model
model.active_path = nil
```

... **does this satisfy this feature request issue ?** (Ie, can it be closed?)

Or do you still think you need a click method for menu items ?

**Comment by arigo** [comment]
Thank you!  This should do exactly what we need.

About picking context menu items from Ruby, I've opened it to a separate issue #425.  It has lower priority as far as we're concerned.

**Comment by DanRathbun** [comment]
@thomthom Can you add a "Fixed-2020.0" label please.
